api/envoy/service/rate_limit_quota/v3/rlqs.proto

```diff
@@ -0,0 +1,178 @@
+// the instance is not sending usage reports. It's on the RLQS server to decide when the instance
+// didn't send reports for too long, and issue an assignment with ``AbandonAction``, telling the
+// client it will no longer send updates for the ``BucketId``.
+//
+// Refer to Rate Limit Quota :ref:`configuration overview <config_http_filters_rate_limit_quota>`
+// for the details.
+
+// Defines the Rate Limit Quota Service (RLQS).
+service RateLimitQuotaService {
+  // Main communication channel: the data plane sends usage reports to the RLQS server,
+  // and the server asynchronously responding with the assignments.
+  rpc StreamRateLimitQuotas(stream RateLimitQuotaUsageReports)
+      returns (stream RateLimitQuotaResponse) {
+  }
+}
+
+message RateLimitQuotaUsageReports {
+  // The usage report for a bucket.
+  //
+  // .. note::
+  //   Note that the first report sent for a ``BucketId`` indicates to the RLQS server that
+  //   the RLQS client is subscribing for the future assignments for this ``BucketId``.
+  //
+  // [#next-free-field: 6]
+  message BucketQuotaUsage {
+    // ``BucketId`` for which request quota usage is reported.
+    BucketId bucket_id = 1 [(validate.rules).message = {required: true}];
+
+    // Time elapsed since the last report.
+    google.protobuf.Duration time_elapsed = 2 [(validate.rules).duration = {
+      required: true
+      gt {}
+    }];
+
+    // Requests the data plane has allowed through.
+    uint64 num_requests_allowed = 3;
+
+    // Requests throttled.
+    uint64 num_requests_denied = 4;
+
+    // Requests in the queue waiting for the assignment.
+    uint64 num_requests_queued = 5;
+  }
+
+  // All quota requests must specify the domain. This enables sharing the quota
+  // server between different applications without fear of overlap.
+  // E.g., "envoy".
+  //
+  // Should only be provided in the first report, all subsequent messages on the same
+  // stream are considered to be in the same domain. In case the domain needs to be
+  // changes, close the stream, and reopen a new one with the different domain.
+  string domain = 1 [(validate.rules).string = {min_len: 1}];
+
+  // An list of quota usage reports. In most cases the order doesn't matter, except the case when
+  // the list contains multiple reports of the same ``BucketId``. Then the earlier report must come
+  // first.
+  repeated BucketQuotaUsage bucket_quota_usages = 2 [(validate.rules).repeated = {min_items: 1}];
+}
+
+message RateLimitQuotaResponse {
+  message BucketAction {
+    // Quota Assignment for the ``BucketId``. Configures the rate limiting strategy and duration.
+    message QuotaAssignmentAction {
+      // A duration after which the assignment is be considered expired.
+      //
+      // If unset, the assignment has no expiration date.
+      //
+      // .. attention::
+      //   Zero-second ``Duration`` expires **immediately**, while unset ``Duration``
+      //   **has no expiration date**.
```

Comment:
This allows RLQS server to force a bucket into `ExpiredAssignmentBehavior`, where Control Plan can set

- a `ExpiredAssignmentBehavior.fallback_rate_limit` (different from  `NoAssignmentBehavior.fallback_rate_limit`), or
- `ExpiredAssignmentBehavior.reuse_last_assignment` - which allows to extend last good assignment before falling back to `NoAssignmentBehavior`

I don't know if this will actually be used. Maybe when RLQS server goes in a reboot, it could proactively tell all RLQS clients fallback to preconfigured `ExpiredAssignmentBehavior`.
I think it still makes sense from the perspective of protocol completeness.

I'll add a note explaining why one would set zero-second assignment TTL.

Reply:
Expanded on the zero-second assignement in https://github.com/envoyproxy/envoy/pull/19793/commits/f7665e145b75fb886c124ed6beaa1df35cd9fd8a.